omisego-sdk/src/main/java/co/omisego/omisego/custom/retrofit2/adapter/OMGCallbackWrapper.kt

```diff
@@ -17,35 +17,47 @@ import retrofit2.Call
 import retrofit2.Callback
 import retrofit2.Response
 import java.net.HttpURLConnection
+import java.util.concurrent.Executor
 
-internal class OMGCallbackWrapper<T>(private val callback: OMGCallback<T>) : Callback<OMGResponse<T>> {
+internal class OMGCallbackWrapper<T>(private val callback: OMGCallback<T>, private val callbackExecutor: Executor?) : Callback<OMGResponse<T>> {
     override fun onFailure(call: Call<OMGResponse<T>>, t: Throwable) {
         when (t) {
             is OMGAPIErrorException -> {
-                callback.fail(t.response)
+                callbackExecutor?.execute { callback.fail(t.response) }
```

Comment:
This is a source of bugs, if I understand that 
```kotlin
retrofit.callbackExecutor()
```
can return null, but the docs says `This may be {@code null}, in which case callbacks should be made synchronously on the background thread.`

Here you're dropping the callback if it's null.
Instead of doing this, I suggest the following:
- Set `Executor` non-null everywhere
- In the `OMGCallAdapterFactory`, get it from retrofit, and if null, use an executor like you did in test that'll just do it in current thread (create a `SynchronousExecutor` or something like that 😄 . 

Doing this will simplify the code by removing nullable, and make it transparent everywhere 😄

Reply:
The reason why I use nullable Executor is the test will fail if it runs through the `MainThreadExecutor` initialization code (because of `Looper.getMainLooper` can't be mocked)

Actually, `callbackExecutor` shouldn't be null because it will always initialize here if it is null

```kotlin
// EWalletClient.kt
/* Initialize the callback executor */
     if (callbackExecutor == null)
         callbackExecutor = MainThreadExecutor()
```

But, the above code can be removed and using your suggested code like this

```kotlin
/* Create retrofit with OMGConverter and OMGCaller */
   eWalletClient.retrofit = Retrofit.Builder().apply {
       //...
       callbackExecutor(callbackExecutor ?: MainThreadExecutor())
  }.build()
```
Okay, I will do it make more sense in the next commit 😄

Reply:
Done

Reply:
Android dependencies can be "mocked", you need to use **Robolectric** to be able to test a code where you need to run Android framework code. (see http://robolectric.org/ ) 😄